fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -958,25 +1076,42 @@ private CompletableFuture<Integer> moveDocsFromPartition(@Nonnull final LucenePa
                     }
                 });
 
-                // update current partition's meta
+                // reset partition info
                 state.context.ensureActive().clear(partitionMetadataKeyFromPartitioningValue(groupingKey, getPartitionKey(partitionInfo)));
-                LucenePartitionInfoProto.LucenePartitionInfo.Builder builder = partitionInfo.toBuilder()
-                        .setCount(partitionInfo.getCount() - records.size())
-                        .setFrom(ByteString.copyFrom(newBoundaryPartitionKey.pack()));
-                savePartitionMetadata(groupingKey, builder);
+                if (!repartitioningContext.emptyingPartition) {
+                    // update current partition's meta
+                    LucenePartitionInfoProto.LucenePartitionInfo.Builder builder = partitionInfo.toBuilder()
+                            .setCount(partitionInfo.getCount() - records.size());
+                    if (repartitioningContext.removingOldest) {
+                        builder.setFrom(ByteString.copyFrom(Objects.requireNonNull(newBoundaryPartitionKey).pack()));
+                    } else {
+                        builder.setTo(ByteString.copyFrom(Objects.requireNonNull(newBoundaryPartitionKey).pack()));
+                    }
+                    savePartitionMetadata(groupingKey, builder);
+                }
+
                 if (LOGGER.isDebugEnabled()) {
                     LOGGER.debug(repartitionLogMessage("Repartitoned records", groupingKey, records.size(), partitionInfo));
                 }
+
                 // value of the "destination" partition's `from` value
                 final Tuple overflowPartitioningKey = toPartitionKey(records.get(0));
-                return findPartitionInfo(groupingKey, overflowPartitioningKey).thenCompose(previousPartition -> {
-                    if (previousPartition == null || previousPartition.getCount() + records.size() > indexPartitionHighWatermark || previousPartition.getId() == partitionInfo.getId()) {
+                final CompletableFuture<LucenePartitionInfoProto.LucenePartitionInfo> futureDestination;
+                if (repartitioningContext.removingOldest) {
+                    futureDestination = findPartitionInfo(groupingKey, overflowPartitioningKey);
+                } else {
+                    futureDestination = getNextNewerPartitionInfo(state.context, groupingKey, overflowPartitioningKey, state.indexSubspace);
+                }
```

Comment:
Well above this we get all the partition info and put it in a list, would it make sense to sort that list, and use that instead of using these methods?
The downside is that you would need two versions of the code for `findPartitionInfo` and `getNextNewerPartitionInfo` depending on whether you're hitting FDB or a list, but it might be a little bit cleaner/faster.

Reply:
in addition,

1. we would need to pass the partition info list down to this method (in the `repartitioningContext`?
2. make sure while implementing the new versions of `findPartitionInfo` and `getNextNewerPartitionInfo` to be aware whether the current partition has been cleared in the DB (a few lines above here).

is the gain in speed is worth this added complexity?